src/odemis/odemisd/main.py

```diff
@@ -328,71 +329,98 @@ def _terminate_all_alive(self):
         It also stops the containers, once no component is running in them.
         """
         mic = self._instantiator.microscope
-        alive = set(mic.alive.value)
-
-        # create a "graph" child->parents
-        parents = {c: set() for c in alive}  # comp -> set of comps (all its parents)
-        for comp in alive:
-            # TODO: if a component was created by delegation, use its creator
-            # instead of itself as "used" by the other components, to ensure
-            # they might not be terminated too early.
+
+        # create a "graph" dependency->parents
+        self._parents = {c: set() for c in set(mic.alive.value)}  # comp -> set of comps (all the comps that depend on it)
+        children_comps = []
+        for comp in self._parents:
             try:
-                for child in comp.children.value:
-                    # TODO: do not add the creator
-                    parents[child].add(comp)
+                for dep in comp.dependencies.value:
+                    # if a component was created by delegation, use its creator
+                    # instead of itself as "used" by the other components, to ensure
+                    # they might not be terminated too early.
+                    for c in self._parents:
+                        if dep in c.children.value:
+                            dep = c
+                            break
+
+                    self._parents[dep].add(comp)
             except Exception:
                 logging.warning("Failed to find children of component %s when terminating",
                                 comp.name, exc_info=True)
 
-        # TODO: for the components created by delegation, either terminate them
-        # before their parents, or don't do it at all (as the parent normally
-        # is in charge of it).
+            for c in self._parents:
```

Comment:
I think it should be possible to remove the children in just one pass, after creating the complete "dependents". Loop through each component, and remove all its children from "dependents". You just need to be careful to not modify the set you loop through, so you need to do a copy:
```
for comp in list(self._dependents.keys()):
    for child in comp.children.value:
        self._dependents.pop(child, None)
```